micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateMetrics.java

```diff
@@ -250,4 +264,97 @@ private SessionFactory unwrap(EntityManagerFactory entityManagerFactory) {
+            EventListenerRegistry registry = ((SessionFactoryImplementor) sessionFactory).getServiceRegistry().getService(EventListenerRegistry.class);
+            MetricsEventHandler metricsEventHandler = new MetricsEventHandler(meterRegistry);
+            registry.appendListeners(EventType.POST_LOAD, metricsEventHandler);
+        }
+
+    }
+
+    class MetricsEventHandler implements PostLoadEventListener {
+
+        private final MeterRegistry meterRegistry;
+
+        MetricsEventHandler(MeterRegistry meterRegistry) {
+            this.meterRegistry = meterRegistry;
+        }
+
+        @Override
+        public void onPostLoad(PostLoadEvent event) {
+            registerQueryMetric(event.getSession().getFactory().getStatistics());
+        }
+
+        void registerQueryMetric(Statistics statistics) {
+            for (String query : statistics.getQueries()) {
+                QueryStatistics queryStatistics = statistics.getQueryStatistics(query);
+                if (Objects.nonNull(queryStatistics)) {
+                    String queryName = query.replace(" ", "_").toLowerCase();
+                    if (Objects.isNull(Search.in(meterRegistry).tags("query", queryName).functionCounter())) {
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheHitCount)
+                                .tags(tags)
+                                .tags("result", "hit", "query", queryName)
+                                .description("The number of query cache hits")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("result", "miss", "query", queryName)
+                                .description("The number of query cache miss")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.puts", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("The number of putting the query into cache")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.executions", queryStatistics, QueryStatistics::getExecutionCount)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("Total number of query executions")
+                                .register(meterRegistry);
+
+                        TimeGauge.builder("hibernate.query.executions.total", queryStatistics, TimeUnit.MILLISECONDS, QueryStatistics::getExecutionTotalTime)
```

Comment:
Can you combine `hibernate.query.executions.total` and `hibernate.query.executions` into a `FunctionTimer`?

Reply:
Yup, sure will adjust